Add protection against huge number of cells

mpuccio opened the pull request:
Hi @shahor02,

still to be tested locally (running it now...), but you can get the gist.
Let me know if you have better ideas.

Cheers,
Max﻿

mpuccio committed:
Add protection against huge number of tracklets/cells

mpuccio commented:
OK, locally it does not break anything on a single TF, I will now try to reconstruct several. We should just bear in mind that the limit I put (factor 10) it is not tuned, we could have a look only at the timings and not at the number of artefacts so I put something out of gut feeling. The parameter is tunable with the command line interface in case it is not effective (or it is disruptive) during data taking.

shahor02 commented on Detectors/ITSMFT/ITS/tracking/src/Tracker.cxx:
```diff
@@ -64,12 +65,21 @@ void Tracker::clustersToTracks(std::function<void(std::string s)> logger, std::f
       break;
     }
+    cellsPerCluster = mTimeFrame->getNumberOfClusters() > 0 ? float(mTimeFrame->getNumberOfCells()) / mTimeFrame->getNumberOfClusters() : 0.f;
+    if (mNumberOfRuns > 9 && cellsPerCluster > mTrkParams[iteration].CellsPerClusterLimit * mAverageCellsPerClusters) {
```
Is clustersToTracks call meant per whole TF (with possibly preselected ROFs for reject)? In PbPb we can have quite large variation of multiplicity, which may also change the non-linear cells/clusters. Should not we have a cut on the abs. fraction of cells/cluster (then, if you want, also very relaxed cut of its factor wrt average)? Note that this protection will not be effective if the anomalous TF happens to the within first 9 iterations.

Also, if I get it right, the iteration is done with a particular set of cuts after suppressing the clusters of the previous iteration, right? Then each level iteration may have quite different cellsPerCluster. Should not per-iteration-type mAverageCellsPerClusters be calculated?
Finally, you apply the cut on the cell level, but already for the trackleting the time jumps by factor 100. Should we apply cut on this level?

mpuccio commented:
Hi @shahor02, I made the cuts "absolute" and added the string output necessary to tune it in case it does not work out of the box (right now I put 10 times the ratios I see in a local pp simulation). I added the selections also at the level of tracklets.

shahor02 commented:
> right now I put 10 times the ratios I see in a local pp simulation

Thanks, pp simulations at which rate? Would not it affect synthetic pbpb runs? Do have numbers for e.g. 400 kHz pp and 50 kHz pbpb?

mpuccio commented:
No, I don't have those number as I don't have such simulations, the local sim is pp@50kHz. However with the new printout we can set easily the threshold.
Note that this will not make any run fail, we would just get errors from ITS and the tracking would not run.

mpuccio commented:
OK, I tried locally with 500kHz and the values put there are in line with those of the previous simulation. I updated the values since I miscalculated them... now the tolerance is factor 20 wrt 500kHz pp. I will simulate some Pb-Pb events and check also in that case.

mpuccio committed:
Update selections and cover cosmics case
